src/radix.zig

```diff
@@ -1,15 +1,58 @@
 const std = @import("std");
 const mem = std.mem;
 
+const SortOrder = enum {
+    ascending,
+    descending,
+};
+
+const SortOptions = struct {
+    order: SortOrder = .ascending,
+    _byte_order: SortOrder = .ascending,
+    key_field: ?@Type(.EnumLiteral) = null,
+    fn canonicalize(comptime self: SortOptions) SortOptions {
+        return .{
+            .order = self.order,
+            ._byte_order = self.order,
+            .key_field = self.key_field,
+        };
+    }
+    fn reversed(comptime self: SortOptions) SortOptions {
+        return .{
+            .order = self.order,
+            ._byte_order = switch (self.order) {
+                .ascending => .descending,
+                .descending => .ascending,
+            },
+            .key_field = self.key_field,
+        };
+    }
+};
+
 const INSSORT_CUTOFF = 55;
 
-inline fn insSort(comptime T: type, noalias array: [*]T, len: usize) void {
+inline fn lt(comptime T: type, comptime options: SortOptions, a: T, b: T) bool {
+    if (options.key_field) |key_field| {
+        const a_key = @field(a, @tagName(key_field));
+        const b_key = @field(b, @tagName(key_field));
+        return switch (options.order) {
+            .ascending => a_key < b_key,
+            .descending => b_key < a_key,
+        };
+    }
+    return switch (options.order) {
+        .ascending => a < b,
+        .descending => b < a,
+    };
+}
```

Comment:
This function does not compare keys in the same way as the radix part of the sort does when the key type is floating. It needs to handle point signed zero and NaNs.

Reply:
This is correct, of course.

Currently my comment about signed zeros will still be true, because zeros of opposite signs will end up in different buckets in the first level of the radix sort, and because we don't have a codepath to skip that and go straight to insertion sort. We should have that codepath though, so my comment about zeros should become wrong and I should then say "the sorting order of zeros is undefined."

About nans, it's correct that nans will generally mess up a whole region of your array that potentially contains a bunch of other non-nan values, and my original comment about the sorted order of nans vs other values is completely incorrect.